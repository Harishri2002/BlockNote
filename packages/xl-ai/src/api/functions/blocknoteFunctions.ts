export type InsertBlocksOperation<T> = {
  type: "add"; // TODO: rename to insert?
  referenceId: string;
  position: "before" | "after";
  blocks: T[];
};

export type UpdateBlocksOperation<T> = {
  type: "update";
  id: string;
  block: T;
};

export type RemoveBlocksOperation = {
  type: "delete"; // TODO: rename to remove?
  ids: string[];
};

export type BlockNoteOperation<T> =
  | InsertBlocksOperation<T>
  | UpdateBlocksOperation<T>
  | RemoveBlocksOperation;

export type InvalidOrOk<T> =
  | {
      result: "invalid";
      reason: string;
    }
  | { result: "ok"; value: T };

/*

testing the validators / invalidorok

export type Car = {
  type: "car";
  brand: string;
};

export type Bike = {
  type: "bike";
  color: string;
};

const objects: any[] = [
  {
    x: 4,
  },
  { type: "car", brand: "Ford", model: "Mustang" },
  { type: "bike", color: "red" },
];

const validateCar = (obj: any): InvalidOrOk<Car> => {
  if (obj.type !== "car") {
    throw new Error("wrong validator called");
  }

  if (typeof obj.brand !== "string") {
    return {
      result: "invalid",
      reason: "no brand",
    };
  }

  return {
    result: "ok",
    value: obj,
  };
};

const validateBike = (obj: any): InvalidOrOk<Bike> => {
  if (obj.type !== "bike") {
    throw new Error("wrong validator called");
  }

  if (typeof obj.color !== "string") {
    return {
      result: "invalid",
      reason: "no color",
    };
  }

  return {
    result: "ok",
    value: obj,
  };
};

type ValidatorFor<T> = {
  type: string;
  validate: (obj: any) => InvalidOrOk<T>;
};

// Refined validator to work with discriminated unions
type TypedValidator<T extends { type: string }> = {
  type: T["type"];
  validate: (obj: any) => InvalidOrOk<T>;
};

// The code below should be generic and not use Car and Bike
const validate = <T extends { type: string }>(
  obj: any,
  validators: readonly TypedValidator<T>[]
) => {
  const validator = validators.find((v) => v.type === obj.type);
  if (!validator) {
    return {
      result: "invalid",
      reason: "Invalid object type",
    } as InvalidOrOk<T>;
  }
  return validator.validate(obj);
};

// Simplified version
const validateAll = <T extends { type: string }>(
  objects: any[],
  validators: readonly TypedValidator<T>[]
) => {
  return objects.map((obj) => validate<T>(obj, validators));
};

// Define vehicle validators with proper types
const carValidator: TypedValidator<Car> = {
  type: "car",
  validate: validateCar,
};

const bikeValidator: TypedValidator<Bike> = {
  type: "bike",
  validate: validateBike,
};

// Create a union type for readability, but not strictly necessary
type Vehicle = Car | Bike;

// Type helper to extract the validated types
type ExtractValidatedTypes<T> = T extends TypedValidator<infer U> ? U : never;

// Helper function that infers the union type automatically
function createValidatorsForTypes<T extends readonly TypedValidator<any>[]>(
  ...validators: T
): readonly TypedValidator<ExtractValidatedTypes<T[number]>>[] {
  return validators;
}

const validators = [carValidator, bikeValidator];

type S = TypedValidator<ExtractValidatedTypes<(typeof validators)[number]>>;

// Use the helper function - no type parameter needed!
const vehicleValidators = createValidatorsForTypes(carValidator, bikeValidator);

// example usage
const validated = validateAll(objects, vehicleValidators);
const o = validated[0];
if (o.result === "ok") {
  const val = o.value; // val is now of type Vehicle (Car | Bike)
}
*/
