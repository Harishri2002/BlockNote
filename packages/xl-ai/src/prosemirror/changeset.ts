import { BlockNoteEditor, getNodeById, updateBlockTr } from "@blocknote/core";
import { ChangeSet, simplifyChanges } from "prosemirror-changeset";
import { ReplaceStep } from "prosemirror-transform";
import { UpdateBlocksOperation } from "../../functions/blocknoteFunctions.js";
import { getCleanDoc } from "./cleanSuggestions.js";

// TODO: extract changeset code + tests + marks / links / mentions etc

export function updateToReplaceSteps(
  editor: BlockNoteEditor,
  op: UpdateBlocksOperation
) {
  const blockPos = getNodeById(op.id, editor.prosemirrorState.doc)!;

  const cleaned = getCleanDoc(editor);

  let changeset = ChangeSet.create(cleaned.cleanTr.doc);

  updateBlockTr(editor, cleaned.cleanTr, blockPos.posBeforeNode, op.block);

  changeset = changeset.addSteps(
    cleaned.cleanTr.doc,
    cleaned.mapsAfterClean(),
    0
  );

  const steps = [];
  const changes = simplifyChanges(changeset.changes, cleaned.cleanTr.doc);
  for (const step of changes) {
    const replaceStart = cleaned.invertMap.map(step.fromA);
    const replaceEnd = cleaned.invertMap.map(step.toA);

    // replace with empty content or first character
    const replacement = cleaned.cleanTr.doc.slice(step.fromB, step.toB);

    if (replacement.openStart > 0 || replacement.openEnd > 0) {
      throw new Error(
        "Replacement expected not to have openStart or openEnd > 0"
      );
    }

    steps.push(new ReplaceStep(replaceStart, replaceEnd, replacement));
  }

  return steps;
}
