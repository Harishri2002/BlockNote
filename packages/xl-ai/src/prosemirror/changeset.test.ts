import { PartialBlock } from "@blocknote/core";
import { expect, it } from "vitest";
import { UpdateBlocksOperation } from "../api/functions/blocknoteFunctions.js";
import {
  getTestEditor,
  testUpdateOperations,
} from "../testUtil/updates/updateOperations.js";
import { updateToReplaceSteps } from "./changeset.js";

function testUpdate(
  update: UpdateBlocksOperation<PartialBlock<any, any, any>>["block"],
  blockId: string
) {
  const editor = getTestEditor();
  const steps = updateToReplaceSteps(
    editor,
    {
      id: blockId,
      type: "update",
      block: update,
    },
    editor.prosemirrorState.doc
  );

  expect(steps).toMatchSnapshot();

  const tr = editor.prosemirrorState.tr;
  for (const step of steps) {
    const mapped = step.map(tr.mapping);
    if (!mapped) {
      throw new Error("Failed to map step");
    }
    tr.step(mapped);
  }
  editor.dispatch(tr);

  const block = editor.getBlock(blockId)!;
  if (update.type) {
    // eslint-disable-next-line
    expect(block.type).toEqual(update.type);
  }
  if (update.props) {
    // eslint-disable-next-line
    expect(block.props).toMatchObject(update.props);
  }
  if (update.content) {
    // eslint-disable-next-line
    expect(block.content).toEqual(update.content);
  }
}

for (const test of testUpdateOperations) {
  it(`${test.description}`, async () => {
    testUpdate(test.updateOp.block, test.updateOp.id);
  });
}
