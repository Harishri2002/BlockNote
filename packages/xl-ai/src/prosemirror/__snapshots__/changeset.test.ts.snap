// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`drop mark and link 1`] = `
[
  {
    "from": 66,
    "slice": {
      "content": [
        {
          "text": "Bold text. ",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 77,
  },
  {
    "from": 77,
    "slice": {
      "content": [
        {
          "text": "Link.",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 82,
  },
]
`;

exports[`drop mark and link and change text within mark 1`] = `
[
  {
    "from": 52,
    "slice": {
      "content": [
        {
          "text": "Hi",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 57,
  },
  {
    "from": 66,
    "slice": {
      "content": [
        {
          "text": "Bold",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 70,
  },
  {
    "from": 70,
    "slice": {
      "content": [
        {
          "text": " the",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 70,
  },
  {
    "from": 70,
    "slice": {
      "content": [
        {
          "text": " text. ",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 77,
  },
  {
    "from": 77,
    "slice": {
      "content": [
        {
          "text": "Link.",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 82,
  },
]
`;

exports[`plain source block, add mention 1`] = `
[
  {
    "from": 10,
    "slice": {
      "content": [
        {
          "attrs": {
            "user": "Jane Doe",
          },
          "type": "mention",
        },
      ],
    },
    "stepType": "replace",
    "to": 15,
  },
]
`;

exports[`standard update 1`] = `
[
  {
    "from": 10,
    "slice": {
      "content": [
        {
          "text": "updated content",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 16,
  },
]
`;

exports[`styles + ic in source block 1`] = `
[
  {
    "from": 27,
    "slice": {
      "content": [
        {
          "text": "updated content",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 48,
  },
]
`;

exports[`styles + ic in source block, remove mark 1`] = `
[
  {
    "from": 34,
    "slice": {
      "content": [
        {
          "text": "are you doing?",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 48,
  },
]
`;

exports[`styles + ic in source block, remove mention 1`] = `
[
  {
    "from": 25,
    "stepType": "replace",
    "to": 28,
  },
]
`;

exports[`styles + ic in source block, update mention prop 1`] = `
[
  {
    "from": 27,
    "slice": {
      "content": [
        {
          "attrs": {
            "user": "Jane Doe",
          },
          "type": "mention",
        },
      ],
    },
    "stepType": "replace",
    "to": 28,
  },
]
`;

exports[`styles + ic in target block, add mark 1`] = `
[
  {
    "from": 10,
    "slice": {
      "content": [
        {
          "marks": [
            {
              "type": "bold",
            },
          ],
          "text": "world!",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 16,
  },
]
`;

exports[`update block prop 1`] = `
[
  {
    "from": 2,
    "slice": {
      "content": [
        {
          "attrs": {
            "textAlignment": "right",
          },
          "type": "paragraph",
        },
      ],
      "openEnd": 1,
    },
    "stepType": "replace",
    "structure": true,
    "to": 3,
  },
]
`;

exports[`update block prop and content 1`] = `
[
  {
    "from": 2,
    "slice": {
      "content": [
        {
          "attrs": {
            "textAlignment": "right",
          },
          "type": "paragraph",
        },
      ],
      "openEnd": 1,
    },
    "stepType": "replace",
    "structure": true,
    "to": 3,
  },
  {
    "from": 3,
    "slice": {
      "content": [
        {
          "text": "What's up",
          "type": "text",
        },
      ],
    },
    "stepType": "replace",
    "to": 8,
  },
]
`;

exports[`update block type 1`] = `
[
  {
    "from": 2,
    "slice": {
      "content": [
        {
          "attrs": {
            "level": 1,
            "textAlignment": "left",
          },
          "type": "heading",
        },
      ],
      "openEnd": 1,
    },
    "stepType": "replace",
    "to": 3,
  },
]
`;

exports[`update block type and content 1`] = `
[
  {
    "from": 2,
    "slice": {
      "content": [
        {
          "attrs": {
            "level": 1,
            "textAlignment": "left",
          },
          "content": [
            {
              "text": "What's up",
              "type": "text",
            },
          ],
          "type": "heading",
        },
      ],
      "openEnd": 1,
    },
    "stepType": "replace",
    "to": 8,
  },
]
`;
